omnigibson/robots/franka_allegro.py

```diff
@@ -228,3 +236,124 @@ def disabled_collision_pairs(self):
+    
+    def remap_thumb_to_allegro(self, coord: np.ndarray) -> np.ndarray:
+        """
+        remap VR thumb tracking data to allegro thumb, based on the bound of the two embodiments
+        Args:
+            bound (np.ndarray) 3D coord of thumb tracking position in the base_link frame
+        """
+        hand_bound = np.array([[0.005, 0.095], [-0.026, 0.104], [-0.005, 0.053]])    # bound for hand
+        allegro_bound = np.array([[0.017, 0.117], [-0.037, 0.132], [-0.096, -0.009]])    # bound for allegro 
+        for i in range(3):
+            coord[i] = np.interp(coord[i], hand_bound[i], allegro_bound[i])
+        return coord
+
+
+    def gen_action_from_vr_data(self, vr_data: dict):
+        action = np.zeros(22)
+        if "hand_data" in vr_data:
+            hand_data = vr_data["hand_data"]
+            if "right" in hand_data and "raw" in hand_data["right"]:
+                # The center of allegro hand lies somewhere between palm and middle proximal
+                target_pos = (hand_data["right"]["raw"]["pos"][0] + hand_data["right"]["raw"]["pos"][12]) / 2
+                target_orn = T.quat_multiply(hand_data["right"]["raw"]["orn"][0], self.vr_rotation_offset[self.default_arm])
+                cur_robot_eef_pos, cur_robot_eef_orn = self.links[self.eef_link_names[self.default_arm]].get_position_orientation()
+                base_pos, base_orn = self.get_position_orientation()
+                rel_target_pos, rel_target_orn = T.relative_pose_transform(target_pos + [0.06, 0, 0], target_orn, base_pos, base_orn)
+                rel_cur_pos = T.relative_pose_transform(cur_robot_eef_pos, cur_robot_eef_orn, base_pos, base_orn)[0]
+                action[:6] = np.r_[rel_target_pos - rel_cur_pos, T.quat2axisangle(rel_target_orn)]
+                # joint order: index, middle, pinky
+                angles = hand_data["right"]["angles"]
+                for f_idx in range(3):
+                    for j_idx in range(3):
+                        action[11 + f_idx * 4 + j_idx] = angles[f_idx + 1][j_idx]
+                # specifically, use ik for thumb
+                thumb_pos, thumb_orn = hand_data["right"]["raw"]["pos"][5], hand_data["right"]["raw"]["orn"][5]
+                local_thumb_pos, local_thumb_orn = T.relative_pose_transform(thumb_pos, thumb_orn, target_pos, target_orn)
+                local_thumb_pos = self.remap_thumb_to_allegro(coord=local_thumb_pos)
+                target_thumb_pos = T.pose_transform(
+                    cur_robot_eef_pos, cur_robot_eef_orn, local_thumb_pos, local_thumb_orn,
+                )[0]
+                action[6: 10] = self.allegro_ik_controller.solve({"thumb": target_thumb_pos})["thumb"]
+
+        else:
+            action_from_controller = super().gen_action_from_vr_data(vr_data)
+            action[:6] = action_from_controller[:6]
+            action[6:] = action_from_controller[6]
+        return action
+
+
+class AllegroIKController:
```

Comment:
hmmm it's a bit weird having a controller directly in the franka class, but I can understand why this is needed. Not sure if there's a more pythonic way to do this, @ChengshuLi @cgokmen ?

Reply:
I think I'll need to integrate [leap hand](http://www.leaphand.com/) in the near future. Perhaps I could setup a DexHandController class so that it can be used for all dex hand?